api/v1alpha1/envoyproxy_types.go

```diff
@@ -86,8 +86,57 @@ type EnvoyProxySpec struct {
 	//
 	// +optional
 	Shutdown *ShutdownConfig `json:"shutdown,omitempty"`
+
+	// FilterPriority defines the order of filters in the Envoy proxy's HTTP filter chain.
+	//
+	// +optional
+	FilterPriority []FilterPriority `json:"filterPriority,omitempty"`
```

Comment:
We need to clarify the API: as I read it, it's all-or-nothing (you either set a filter order, in which case you need to set _every filter you want executed_, or you get the default) -- is that correct?

I think that's a reasonable API, but it makes simple things ("add a WASM filter before authn, please") as hard as hard things ("totally re-order how policy is applied, including my WASM, ext_proc, ext_authz, and rate limiting"). It's also very copy+paste heavy: you look up the official order and copy+paste it to start editing, and next time you copy+paste your edited version for the new one, and so on. This winds up making it more expensive to iterate on APIs as a project.

Instead it might be easier to have a split API, along the lines of:

```suggestion
        type FilterOrder struct {
                Add    []FilterPriority `json:"add,omitempty"`
                Custom []FilterPriority `json:"custom,omitempty"`
        }
        FilterOrder FilterOrder `json:"filterOrder,omitempty"`
```

Then for small changes a user can `add`, or for total control they can use `replace`:

```yaml
filterOrder:
  add:
    - filter: envoy.filters.http.wasm
      priority: 10
```
```yaml
filterOrder:
  custom:
    - filter: envoy.filters.http.cors
      priority: 901
    - filter: envoy.filters.http.ext_authz
      priority: 801
    - filter: envoy.filters.http.basic_authn
      priority: 701
    - filter: envoy.filters.http.oauth2
      priority: 601
    - filter: envoy.filters.http.jwt_authn
      priority: 501
    - filter: envoy.filters.http.fault
      priority: 401
    - filter: envoy.filters.http.local_ratelimit
      priority: 301
    - filter: envoy.filters.http.ratelimit
      priority: 201
```

And of course we'd expect the two to work together as well:
```yaml
# these are the same
filterOrder:
  add:
    - filter: envoy.filters.http.wasm
      priority: 10
  custom:
    - filter: envoy.filters.http.cors
      priority: 901
    - filter: envoy.filters.http.ext_authz
      priority: 801
    - filter: envoy.filters.http.basic_authn
      priority: 701
    - filter: envoy.filters.http.oauth2
      priority: 601
    - filter: envoy.filters.http.jwt_authn
      priority: 501
    - filter: envoy.filters.http.fault
      priority: 401
    - filter: envoy.filters.http.local_ratelimit
      priority: 301
    - filter: envoy.filters.http.ratelimit
      priority: 201
---
filterOrder:
  custom:
    - filter: envoy.filters.http.cors
      priority: 901
    - filter: envoy.filters.http.ext_authz
      priority: 801
    - filter: envoy.filters.http.basic_authn
      priority: 701
    - filter: envoy.filters.http.oauth2
      priority: 601
    - filter: envoy.filters.http.jwt_authn
      priority: 501
    - filter: envoy.filters.http.fault
      priority: 401
    - filter: envoy.filters.http.local_ratelimit
      priority: 301
    - filter: envoy.filters.http.ratelimit
      priority: 201
    - filter: envoy.filters.http.wasm
      priority: 10
```

Reply:
@ZackButcher It doesn't have to be a full list.

Each filters have a default priority, as the comment shows:
// - envoy.filters.http.cors               0
//
// - envoy.filters.http.ext_authz          100
//
// - envoy.filters.http.basic_authn        200
//
// - envoy.filters.http.oauth2             300
//
// - envoy.filters.http.jwt_authn          400
//
// - envoy.filters.http.fault              500
//
// - envoy.filters.http.local_ratelimit    600
//
// - envoy.filters.http.ratelimit          700


If we want to change the default priority of the oauth2 filter to place it between cors and ext_auth, it can be done with the following config:

```
filterPriority:
  envoy.filters.http.oauth2: 50
```

For custom extensions like wasm, lua, and extproc, we plan to add a priority field in their individual spec, for exampel:

```
  wasm:
    - name: my_plugin_1
      priority: 450
      code:
        image:  # the oci image that contains wasm code
          url: <oci-registry>.foo.bar.com/my_plugin_1:v1 
          pullSecret: wasmPullSecret
      ...
    - name: my_plugin_2
      priority: 250
      code:
        image:  # the oci image that contains wasm code
          url: <oci-registry>.foo.bar.com/my_plugin_2:v1 
          pullSecret: wasmPullSecret
      ...      

```  

my_plugin_1 will be placed between jwt_authn and fault.
my_plugin_2 will be placed between basic_authn and jwt_authn.

The ultimate order of these filters are: cors(0),oauth2(50),basic_authn(200), wasm_my_plugin_2(250),jwt_authn(400), wasm_my_plugin_1(450),fault(500),local_ratelimit(600),ratelimit(700)